iwyu_output.cc

```diff
@@ -726,16 +727,16 @@ bool DeclCanBeForwardDeclared(const Decl* decl, string* reason) {
 
   if (isa<ClassTemplateDecl>(decl)) {
     // Class templates can always be forward-declared.
-  } else if (const auto* record = dyn_cast<RecordDecl>(decl)) {
-    // Record decls can be forward-declared unless they don't have
+  } else if (const auto* tag_decl = dyn_cast<TagDecl>(decl)) {
+    // Tag decls can be forward-declared unless they don't have
     // a type name to forward-declare (that includes lambdas).
-    if (!record->getIdentifier()) {
+    if (!tag_decl->getIdentifier()) {
       *reason = "declaration has no name";
       return false;
     }
   } else {
     // Other decl types are not forward-declarable.
-    *reason = "not a record or class template";
+    *reason = "not a tag type or class template";
```

Comment:
Maybe, it would be better to explicitly enumerate allowed declaration types: `"not a record, enumeration or class template"`?
Or `"not a class, struct, union, enum or class template"`?

Reply:
It's a little noisy. I was hoping to teach `DeclCanBeForwardDeclared` and friends to consider functions forward-declarable. Every construct we add support for would have to patch this line to add `... or X`.

But it's also nice to be explicit. I think your first suggestion is OK -- `"not a record, enumeration or class template"` -- let's use that.